File: app/src/main/java/org/mesonet/app/radar/RadarSiteDataProvider.java
package org.mesonet.app.radar;


import android.app.Activity;
import android.content.Context;
import android.location.Location;
import android.util.Pair;
import android.util.Xml;

import org.mesonet.app.BasicViewHolder;
import org.mesonet.app.dependencyinjection.PerFragment;
import org.mesonet.app.filterlist.FilterListFragment;
import org.mesonet.app.site.SiteSelectionInterfaces;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Observable;

import javax.inject.Inject;


@PerFragment
public class RadarSiteDataProvider extends Observable implements FilterListFragment.FilterListDataProvider, SiteSelectionInterfaces.SelectSiteListener
{
    private String mSelectedRadar = "KTLX";
    Map<String, RadarModel.RadarDetailModel> mRadarDetail;



    @Inject
    public RadarSiteDataProvider(Activity inContext)
    {
        InputStream radarList = inContext.getResources().openRawResource(inContext.getResources().getIdentifier("radar_list", "raw", inContext.getPackageName()));
        mRadarDetail = ProcessRadarXml(radarList);
        setChanged();
        notifyObservers();
    }



    private Map<String, RadarModel.RadarDetailModel> ProcessRadarXml(InputStream inRadarXmlStream)
    {
        XmlPullParser xmlParser = Xml.newPullParser();

        Map<String, RadarModel.RadarDetailModel> result = null;

        try {
            xmlParser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            xmlParser.setInput(inRadarXmlStream, null);
            result = ParseRadarXml(xmlParser);
        }
        catch (XmlPullParserException | IOException e)
        {
            e.printStackTrace();
        }

        try {
            inRadarXmlStream.close();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return result;
    }



    private Map<String, RadarModel.RadarDetailModel> ParseRadarXml(XmlPullParser inXmlParser) throws XmlPullParserException, IOException
    {
        Map<String, RadarModel.RadarDetailModel> result = new HashMap<>();

        while (inXmlParser.next() != XmlPullParser.END_TAG)
        {
            if (inXmlParser.getEventType() != XmlPullParser.START_TAG)
                continue;

            String name = inXmlParser.getName();
            // Starts by looking for the entry tag
            if (name.equals("key"))
            {
                if(inXmlParser.next() == XmlPullParser.TEXT)
                {
                    String key = inXmlParser.getText();

                    inXmlParser.next();
                    inXmlParser.next();
                    inXmlParser.next();

                    if(inXmlParser.getName().equals("dict"))
                        result.put(key, ParseRadarDetail(inXmlParser));
                }
            }
        }

        return result;
    }



    private RadarModel.RadarDetailModel ParseRadarDetail(XmlPullParser inXmlParser) throws XmlPullParserException, IOException
    {
        RadarModel.RadarDetailModel.Builder builder = new RadarModel.RadarDetailModel.Builder();

        while (inXmlParser.next() != XmlPullParser.END_TAG || (inXmlParser.getName() == null || !inXmlParser.getName().equals("dict")))
        {
            if (inXmlParser.getEventType() != XmlPullParser.START_TAG)
                continue;

            if(inXmlParser.getName().equals("key"))
            {
                inXmlParser.next();

                String key = inXmlParser.getText();

                inXmlParser.next();
                inXmlParser.next();
                inXmlParser.next();

                String type = inXmlParser.getName();

                inXmlParser.next();

                String value = inXmlParser.getText();

                builder.SetValue(key, type, value);
            }
        }

        return builder.Build();
    }



    public RadarModel.RadarDetailModel GetRadarDetail(String inKey)
    {
        return mRadarDetail.get(inKey);
    }



    public RadarModel.RadarDetailModel GetRadarDetail()
    {
        return GetRadarDetail(mSelectedRadar);
    }



    public String GetRadarName()
    {
        return mSelectedRadar + " - " + mRadarDetail.get(mSelectedRadar).GetName();
    }

    @Override
    public Map<String, BasicViewHolder.BasicViewHolderData> AllViewHolderData() {
        Map<String, BasicViewHolder.BasicViewHolderData> data = new LinkedHashMap<>();

        List<String> keys = new ArrayList<>(mRadarDetail.keySet());

        for(int i = 0; i < keys.size(); i++)
        {
            RadarModel.RadarDetailModel radar = mRadarDetail.get(keys.get(i));

            Location location = new Location("none");
            location.setLatitude(radar.GetLatitude());
            location.setLongitude(radar.GetLongitude());

            data.put(keys.get(i), new BasicViewHolder.BasicViewHolderData(keys.get(i) + " - " + mRadarDetail.get(keys.get(i)).GetName(), location, false));
        }

        return data;
    }

    @Override
    public String CurrentSelection() {
        return mSelectedRadar;
    }

    @Override
    public Observable GetDataObservable() {
        return this;
    }

    @Override
    public void SetResult(String inResult) {
        mSelectedRadar = inResult;
        setChanged();
        notifyObservers();
    }
}

File: app/src/main/java/org/mesonet/app/radar/RadarSiteDataProvider.kt
package org.mesonet.app.radar


import android.app.Activity
import android.location.Location
import android.util.Xml

import org.mesonet.app.BasicViewHolder
import org.mesonet.app.dependencyinjection.PerFragment
import org.mesonet.app.filterlist.FilterListFragment
import org.mesonet.app.site.SiteSelectionInterfaces
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException

import java.io.IOException
import java.io.InputStream
import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedHashMap
import java.util.Observable

import javax.inject.Inject


@PerFragment
class RadarSiteDataProvider @Inject
constructor(inContext: Activity) : Observable(), FilterListFragment.FilterListDataProvider, SiteSelectionInterfaces.SelectSiteListener {
    private var mSelectedRadar = "KTLX"
    internal var mRadarDetail: Map<String, RadarModel.RadarDetailModel>? = null


    init {
        val radarList = inContext.resources.openRawResource(inContext.resources.getIdentifier("radar_list", "raw", inContext.packageName))
        mRadarDetail = ProcessRadarXml(radarList)
        setChanged()
        notifyObservers()
    }


    private fun ProcessRadarXml(inRadarXmlStream: InputStream): Map<String, RadarModel.RadarDetailModel>? {
        val xmlParser = Xml.newPullParser()

        var result: Map<String, RadarModel.RadarDetailModel>? = null

        try {
            xmlParser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
            xmlParser.setInput(inRadarXmlStream, null)
            result = ParseRadarXml(xmlParser)
        } catch (e: XmlPullParserException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        try {
            inRadarXmlStream.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return result
    }


    @Throws(XmlPullParserException::class, IOException::class)
    private fun ParseRadarXml(inXmlParser: XmlPullParser): Map<String, RadarModel.RadarDetailModel> {
        val result = HashMap<String, RadarModel.RadarDetailModel>()

        while (inXmlParser.next() != XmlPullParser.END_TAG) {
            if (inXmlParser.eventType != XmlPullParser.START_TAG)
                continue

            val name = inXmlParser.name
            // Starts by looking for the entry tag
            if (name == "key") {
                if (inXmlParser.next() == XmlPullParser.TEXT) {
                    val key = inXmlParser.text

                    inXmlParser.next()
                    inXmlParser.next()
                    inXmlParser.next()

                    if (inXmlParser.name == "dict")
                        result[key] = ParseRadarDetail(inXmlParser)
                }
            }
        }

        return result
    }


    @Throws(XmlPullParserException::class, IOException::class)
    private fun ParseRadarDetail(inXmlParser: XmlPullParser): RadarModel.RadarDetailModel {
        val builder = RadarModel.RadarDetailModel.Builder()

        while (inXmlParser.next() != XmlPullParser.END_TAG || inXmlParser.name == null || inXmlParser.name != "dict") {
            if (inXmlParser.eventType != XmlPullParser.START_TAG)
                continue

            if (inXmlParser.name == "key") {
                inXmlParser.next()

                val key = inXmlParser.text

                inXmlParser.next()
                inXmlParser.next()
                inXmlParser.next()

                val type = inXmlParser.name

                inXmlParser.next()

                val value = inXmlParser.text

                builder.SetValue(key, type, value)
            }
        }

        return builder.Build()
    }


    @JvmOverloads
    fun GetRadarDetail(inKey: String = mSelectedRadar): RadarModel.RadarDetailModel {
        return mRadarDetail!![inKey]!!
    }


    fun GetRadarName(): String {
        return mSelectedRadar + " - " + mRadarDetail!![mSelectedRadar]!!.GetName()
    }

    override fun AllViewHolderData(): Map<String, BasicViewHolder.BasicViewHolderData> {
        val data = LinkedHashMap<String, BasicViewHolder.BasicViewHolderData>()

        val keys = ArrayList(mRadarDetail!!.keys)

        for (i in keys.indices) {
            val radar = mRadarDetail!![keys[i]]

            val location = Location("none")
            location.latitude = radar?.GetLatitude()!!.toDouble()
            location.longitude = radar.GetLongitude().toDouble()

            data[keys[i]] = BasicViewHolder.BasicViewHolderData(keys[i] + " - " + mRadarDetail!![keys[i]]!!.GetName(), location, false)
        }

        return data
    }

    override fun CurrentSelection(): String {
        return mSelectedRadar
    }

    override fun GetDataObservable(): Observable {
        return this
    }

    override fun SetResult(inResult: String) {
        mSelectedRadar = inResult
        setChanged()
        notifyObservers()
    }
}
